packages/backend/server/src/plugins/copilot/provider.ts

```diff
@@ -0,0 +1,135 @@
+import assert from 'node:assert';
+
+import { Injectable, Logger } from '@nestjs/common';
+
+import { Config } from '../../fundamentals';
+import {
+  CapabilityToCopilotProvider,
+  CopilotConfig,
+  CopilotProvider,
+  CopilotProviderCapability,
+  CopilotProviderType,
+} from './types';
+
+type CopilotProviderConfig = CopilotConfig[keyof CopilotConfig];
+
+interface CopilotProviderDefinition<C extends CopilotProviderConfig> {
+  // constructor signature
+  new (config: C): CopilotProvider;
+  // type of the provider
+  readonly type: CopilotProviderType;
+  // capabilities of the provider, like text to text, text to image, etc.
+  readonly capabilities: CopilotProviderCapability[];
+  // asserts that the config is valid for this provider
+  assetsConfig(config: C): boolean;
+}
+
+// registered provider factory
+const COPILOT_PROVIDER = new Map<
+  CopilotProviderType,
+  (config: Config, logger: Logger) => CopilotProvider
+>();
+
+// map of capabilities to providers
+const PROVIDER_CAPABILITY_MAP = new Map<
+  CopilotProviderCapability,
+  CopilotProviderType[]
+>();
+
+// config assertions for providers
+const ASSERT_CONFIG = new Map<CopilotProviderType, (config: Config) => void>();
+
+export function registerCopilotProvider<
```

Comment:
can use the same strategy in `/oauth/register.ts` for auto registering providers

Reply:
will adjust it in after pr